src/Oscoin/Crypto/PubKey.hs

```diff
@@ -1,190 +1,104 @@
-    , PrivateKey
-    , serialisePrivateKey
-    , deserialisePrivateKey
+    (
+    -- * Signing and verifying messages
+      HasDigitalSignature(..)
+    , unsign
 
     , KeyPair
-    , generateKeyPair
 
-    , sign
-    , signBytes
-    , signed
-    , unsign
-    , verify
-    , verifyBytes
+    -- * Internals
+    , Signed(..)
+    , PublicKey(..)
+    , PrivateKey(..)
     ) where
 
 import           Oscoin.Prelude
 
 import qualified Oscoin.Crypto.Hash as Crypto
 
-import qualified Crypto.PubKey.ECC.ECDSA as ECDSA
-import           Crypto.PubKey.ECC.Generate (generate)
-import           Crypto.PubKey.ECC.Types (CurveName(SEC_p256k1), getCurveByName)
 import           Crypto.Random.Types (MonadRandom)
 
-import qualified Codec.CBOR.Read as CBOR (deserialiseFromBytes)
-import qualified Codec.CBOR.Write as CBOR (toLazyByteString)
 import           Codec.Serialise (Serialise(..))
-import qualified Codec.Serialise as CBOR
-import qualified Codec.Serialise.Decoding as CBOR
-import qualified Codec.Serialise.Encoding as CBOR
-import           Codec.Serialise.JSON (deserialiseParseJSON, serialiseToJSON)
 import           Codec.Serialise.Orphans ()
-import           Control.Monad.Fail (fail)
 import           Data.Aeson
                  (FromJSON(..), ToJSON(..), object, withObject, (.:), (.=))
-import qualified Data.ByteString.Lazy as LBS
-import qualified Data.Hashable as H
+import           Data.ByteArray (ByteArrayAccess)
 import           Data.Text.Prettyprint.Doc
 
-data PublicKey = PublicKey ECDSA.PublicKey (Crypto.Hashed ECDSA.PublicKey)
-    deriving (Show, Generic)
-
-mkPublicKey :: ECDSA.PublicKey -> PublicKey
-mkPublicKey pk = PublicKey pk (Crypto.hash pk)
-
-instance H.Hashable PublicKey where
-    hashWithSalt salt = H.hashWithSalt salt . Crypto.fromHashed . publicKeyHash
-
-instance Serialise PublicKey where
-    encode (PublicKey pk _) = encode pk
-    decode = mkPublicKey <$> decode
-
-instance ToJSON PublicKey where
-    toJSON = serialiseToJSON
-
-instance FromJSON PublicKey where
-    parseJSON = deserialiseParseJSON
-
-instance Crypto.Hashable PublicKey where
-    hash (PublicKey _ h) = Crypto.toHashed (Crypto.fromHashed h)
-
-instance Eq PublicKey where
-    (==) (PublicKey _ h) (PublicKey _ h') = h == h'
-
-instance Ord PublicKey where
-    (<=) (PublicKey _ h) (PublicKey _ h') = h <= h'
-
-publicKeyHash :: PublicKey -> Crypto.Hashed ECDSA.PublicKey
-publicKeyHash (PublicKey _ h) = h
+{------------------------------------------------------------------------------
+ Types
+------------------------------------------------------------------------------}
 
-newtype PrivateKey = PrivateKey ECDSA.PrivateKey
-    deriving (Show, Eq)
+-- | A 'PublicKey' over a certain crypto 'c' and an algorithm.
+-- /DO NOT/ use this type directly, it's exported only for the sake of
+-- implementing the 'HasDigitalSignature' instance in 'PubKey.RealWorld'.
+-- Use the abstract PK/SK instead.
+data PublicKey c k = PublicKey k (Crypto.Hashed c k)
```

Comment:
Hmm, seems easy to miss the docstring and use this type - any ideas how to prevent that?

Reply:
On top of my head the only way to do that would be to move _both_ the `MockCrypto` and `RealWorldCrypto` implementations into this module, which would become a bit messy, though.

Another "sweet-spot" solution might be to move `PublicKey` and `PrivateKey` inside an `Internal` module, a bit like you did with the SQLite transaction stuff, to stress the fact it shouldn't be used directly. 

However, now we _never_ reference directly a `PublicKey` in the types, so users trying to actually use this type would have GHC complain that it cannot unify `PublicKey c k ~ PK c`. The risk is more on novel code written by contributors not familiar with this abstraction.

Reply:
Moving to internal sounds like a good idea - I'd also think that calling the internal ones PK/SK makes more sense, while PublicKey/PrivateKey are more readable for external usage. Maybe it's worth creating an issue and tagging it with [easy]?

Reply:
cf21232

⚠️ Will open an issue for the renaming as soon as we merge (visual cue for my own benefit).